src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs

```diff
@@ -1011,76 +1025,116 @@ protected bool CheckType(TexlNode node, DType nodeType, DType expectedType, IErr
+        // Coercion is supported and nodeToCoercedTypeMap is updated.
+        //
+        // Many functions return the same type and column name (in the days before ConsistentOneColumnTableResult) as the first argument,
+        // pass context (to get at features) and an out parameter for the type to use as the result of such a function.
+        // If not needed, use one of the overloads without context and the out parameter.
+        // Date/DateTime/Time are special and will not snap to expectedType but will retain the subtype.
+        private bool CheckColumnType(DType type, TexlNode arg, DType expectedType, IErrorContainer errors, ErrorResourceKey errKey, ref Dictionary<TexlNode, DType> nodeToCoercedTypeMap, CheckTypesContext context, out DType returnType)
         {
             Contracts.Assert(type.IsValid);
             Contracts.AssertValue(arg);
             Contracts.Assert(expectedType.IsValid);
-            Contracts.AssertValue(errors);
-
+            Contracts.AssertValue(errors);
+
+            returnType = type;
+
             IEnumerable<TypedName> columns;
             if (!type.IsTable || (columns = type.GetNames(DPath.Root)).Count() != 1)
             {
                 errors.EnsureError(DocumentErrorSeverity.Severe, arg, TexlStrings.ErrInvalidSchemaNeedCol);
                 return false;
             }
-            else
+
+            var column = columns.Single();
+            var colType = column.Type;
+
+            if (!expectedType.Accepts(colType))
             {
-                var column = columns.Single();
-                if (!expectedType.Accepts(column.Type))
+                if (SupportsParamCoercion && column.Type.CoercesTo(expectedType))
+                {
+                    returnType = DType.CreateTable(new TypedName(expectedType, column.Name));
+                    CollectionUtils.Add(ref nodeToCoercedTypeMap, arg, returnType);
+                }
+                else
                 {
-                    if (SupportsParamCoercion && column.Type.CoercesTo(expectedType))
-                    {
-                        expectedType = DType.CreateTable(new TypedName(expectedType, column.Name));
-                        CollectionUtils.Add(ref nodeToCoercedTypeMap, arg, expectedType);
-                    }
-                    else
-                    {
-                        errors.EnsureError(DocumentErrorSeverity.Severe, arg, errKey, column.Name.Value);
-                        return false;
-                    }
+                    errors.EnsureError(DocumentErrorSeverity.Severe, arg, errKey, column.Name.Value);
+                    return false;
                 }
             }
-
+
+            if (context?.Features.ConsistentOneColumnTableResult == true)
+            {
+                // Note that DateTime retains the subtype and does not snap to expectedType.
+                // For example, DateAdd([Date(2000,1,1)],3) should return *[Value:D] and not *[Value:d]

```

Comment:
Can this be handled in DateTime specific case instead of here? For example,
CheckDateColumnType

Reply:
It would be difficult to hoist this out.  ColType is only down in here, and it would need to be either an additional out parameter or recomputed.  And then this code to create the table would need to be repeated in all of the Check*ColumnType functions that take a context.  Seemed to me that the best solution with the least code duplication (and opportunity to get out of sync) was to put it here and comment it well.